docs/RFCS/20171003_inverted_indexes.md

```diff
@@ -0,0 +1,506 @@
+| `x:b`                       | `2`                                                        |
+| `y:7`                       | `1,3`                                                      |
+| `z:a:true`                  | `2`                                                        |
+| `z:b:false`                 | `2,3`                                                      |
+
+The question we posed above `valueCol->z->b = false` is now incredibly natural
+to answer with this type of index. Now it’s a straightforward lookup into the
+KV, just the way ordinary indexes are.
+
+This general approach to indexing is usable on any tokenizable datatype, such
+as arrays, text, and even XML.
+
+# Reference-level explanation
+
+## Detailed Design: Key encoding
+
+Postgres supports two flavors of inverted indexes for JSON `jsonb_ops` and
+`jsonb_path_ops` . The first method, `jsonb_ops`, works by adding every JSON
+key and value from the indexed column into the index as a key. It’s essentially
+no different from full-text search - keys and values are treated almost
+identically, and all of the hierarchical information from the input JSON is
+discarded. Given the following JSON:
+
+    {"x": "b", "z": {"a": true, "b": false}}
+
+the keys produced would be:
+
+    Kx
+    Kb
+    Kz
+    Ka
+    Vtrue
+    Vfalse
+
+To find the results for the following containment query,
+
+    SELECT * from t WHERE jsonCol @> {"x":"b", "z:"{"a":true}}
+
+we would have to look up the index key for every key and value in the input to
+get a set of primary keys per object, then take the intersection of those sets,
+and finally re-check the containment condition for all of the rows in that
+intersection to make sure that the structure is correct. This is inefficient in
+terms of both time and space. In this example, it’s obvious that a very large
+number of rows can have `true` as a value in their JSON column so we would have
+to perform a join on a very large data set. The resultant index will also be
+quite large on disk, since every individual key and value gets indexed.
+
+The second type of JSON index that Postgres supports is called
+`jsonb_path_ops`. This type of index encodes the full JSON path for every leaf
+value into the index key. For the example above, the resultant index keys would
+be the following:
+
+    x:b
+    z:a:true
+    z:b:false
+
+This type of index produces significantly fewer keys, only one per leaf from
+original JSON, not one per token. Performing lookups on this encoding is very
+efficient for full path queries because it’s just one lookup.
+
+From user feedback we’ve received, users are mostly interested in running
+queries supported by the second type. The second type is also more space and
+time efficient. Therefore, we propose implementing the second variety of index
+initially. Once we implement full text search, we will have the first variety
+for free.
```

Comment:
It's not clear to me from reading this how much overlap there is in the implementation between the two flavors. Could you comment on that, since it sounds like we'll eventually want to have both kinds?